### DISCUSSION TITLE
Google Drive Support

### DISCUSSION THREAD

**Discussion by daylightstudio** [spatie/laravel-medialibrary#2380] [Ideas]
First, thanks for creating such a great library!

I'm needing to incorporate Google Drive to store some assets. I have it pretty much working after following these steps and am using a custom PathGenerator class to generate my path as suggested in this closed issue:
https://github.com/spatie/laravel-medialibrary/issues/1331

However, when I add media to the model it is creating an additional folder with the name of the Google Folder ID (not the friendly name). I traced it to the following line:
https://github.com/spatie/laravel-medialibrary/blob/ebbc996db457adecc778db6030d22ef72b495d59/src/MediaCollections/Filesystem.php#L309

This has to do with how directories are created with the adapter and it needing the Google folder id and not the name of the folder (you can actually have folders with the same name on Google Drive because of this difference).Changing that line to something like the following fixes the issue (assuming the drive is named 'google'):
```
if (! in_array($diskDriverName, ['s3', 'google'], true)) {
```

My ultimate fix was to extend this class and overwrite the method and then in a service provider bind it as in the following example, however, it would be great to not have to do this:
```
$this->app->bind(Filesystem::class, FixedFilesystem::class);
```

Thanks again!

**Comment by asim-altayb** [comment]
please give us more details

**Comment by daylightstudio** [comment]
Sure. This requires a little bit of time to setup but I followed these steps to get Google Drive to work with Laravel Storage:
https://github.com/ivanvermeyen/laravel-google-drive-demo
https://github.com/ivanvermeyen/laravel-google-drive-demo/blob/master/routes/web.php (examples of usage line 64 in particular which creates a subfolder)

**Model Code:**
I setup the model [per instructions](https://spatie.be/docs/laravel-medialibrary/v9/basic-usage/preparing-your-model) and added the following registration (note the `useDisk('google')`):
```
public function registerMediaCollections(): void
{
        $this->addMediaCollection('solar_panel_system_files')->useDisk('google');
}
```

**Custom PathGenerator**
I setup a[ custom path generator ](https://spatie.be/docs/laravel-medialibrary/v9/advanced-usage/using-a-custom-directory-structure)to create the folders on Google Drive on upload if they didn't exist which below is a simplified version of what I have (haven't tested this simplified version but you should get the gist):
```
use Storage;
use Spatie\MediaLibrary\MediaCollections\Exceptions\FileDoesNotExist;

class GoogleDrivePathGenerator implements PathGenerator
{
    public function getPath(Media $media): string
    {
        $friendlyPath = 'my/path/file';
        $path = static::fromFriendlyPath($friendlyPath, '/', true);
        config('filesystems.disks.google.folderId', $path);

        return $path.'/';
    }

    public function getPathForResponsiveImages(Media $media): string
    {
        // not relevant for this example
    }

    public function getPathForConversions(Media $media): string
    {
        // not relevant for this example
    }


    public static function fromFriendlyPath($path, $dir = '', $create = false)
    {
        $pathParts = explode(DIRECTORY_SEPARATOR, trim($path, '/'));
        foreach ($pathParts as $i => $part) {
            $p = static::getPathData($part, $dir);
            if (!$p && $create) {
                Storage::cloud()->makeDirectory($dir.'/'.$part);
                $p = static::getPathData($part, $dir);
            }

            if ($p) {
                $dir = $p['path'];
            } else {
                throw FileDoesNotExist::create($path);
            }
        }

        return $dir;
    }

    public static function getPathData($filename, $dir = '/', $type = null)
    {
        $contents = collect(Storage::cloud()->listContents($dir));
        $filtered = $contents->where('name', '=', $filename);

        if ($type) {
            $filtered->where('type', '=', $type);
        }

        return $filtered->first();
    }
}
```

The problem I would run into is that on upload, it would create the folders that I specified and upload the file, however there would also be an additional folder created that would be named the same as the Google Folder ID and I tracked it down to the line of code mentioned above. This is because Google uses the file/folder ID to identify it and not the name of the file/folder (probably similar to S3 which is why it's in that array for exclusion).

**Comment by fdelvals** [comment]
Hi, 
Thank you for your time. I am tryng to implement this solution but I don't think I fully understand it...
So is it working with your tutorial?

**Reply by daylightstudio** [reply]
Hi @fdelvals, let me know what part you are having trouble with (any error messages and what you've tried), and I'll do my best to help.

**Reply by sidahmedabdelillah** [reply]
Hello i tried this sollution but a couldnt really follow it , can please explain more

**Comment by erikn69** [comment]
I'm using with [masbug/flysystem-google-drive-ext](https://github.com/masbug/flysystem-google-drive-ext) wihout problems

**Reply by sidahmedabdelillah** [reply]
I tried it and i still have that problem 
Can you please explain how

**Reply by erikn69** [reply]
> when I add media to the model it is creating an additional folder with the name of the Google Folder ID (not the friendly name)

`masbug/flysystem-google-drive-ext` uses the friendly name, that was the problem, maybe you have another problem, are you setting public permissions? What error do you have?

**Reply by sidahmedabdelillah** [reply]
im having this response 
File not found at path: 10/HI1PuFMHBLUrpK17Ky2KQYIyeMhsyiTqhRtBHrUr.jpg

**Reply by sidahmedabdelillah** [reply]
Should i use a custom path generator ?

**Reply by erikn69** [reply]
>`masbug/flysystem-google-drive-ext` uses the friendly name

No path generator needed,  is `HI1PuFMHBLUrpK17Ky2KQYIyeMhsyiTqhRtBHrUr` google id or this package id?

**Reply by sidahmedabdelillah** [reply]
It is Working now perfectly it was a silly problem from my part i confirm that this masbug/flysystem-google-drive-ext works fine

**Reply by erikn69** [reply]
@sidahmedabdelillah give an ⭐to `masbug/flysystem-google-drive-ext` if works for you

**Reply by booni3** [reply]
Thanks for this. It works... however Google drive is INCREDIBLY slow vs. AWS. Shame as I have 20+TB of unused space on google workspaces!
